### PR TITLE
Added support for reading from DynamoDB Streams

### PR DESCRIPTION
**Pull request by nareshz**
**Things added**
- Support for reading records from DynamoDB Streams
   -  processing of DynamoDB Streams for different tables concurrently using threads
   - retrieval of shards from streams and processing so that order of changes on source database is maintained
   - strategy for processing records within shards handling all the corner cases.
  
- Unit tests for testing individual functions.
  - tests for checking DynamoDB Streams API calls.
  - tests for checking whole reading logic by creating a mock DynamoDB Streams client. 

**Things not handled**
- Writing data read from DynamoDB Streams to Cloud Spanner and changes to reportFile.
- Changes to README will be added after complete streaming migration solution is ready.

- [x] Tests pass

### CODE REVIEW

File: sources/dynamodb/streaming.go [hunk 1]
```diff
@@ -58,3 +68,186 @@ func NewDynamoDBStream(client dynamodbiface.DynamoDBAPI, srcTable string) (strin
...
+}
+
+// ProcessStream processes the latest enabled DynamoDB Stream for a table.
+// It searches for shards within stream and for each shard it creates a
+// seperate working thread to process records within it.
+func ProcessStream(wgStream *sync.WaitGroup, streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, streamInfo *Info, conv *internal.Conv, streamArn, srcTable string) {
+	defer wgStream.Done()
+	wgShard := &sync.WaitGroup{}
+
+	var lastEvaluatedShardId *string = nil
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:90]: Rename it to lastprocessedshardid to avoid confusion

File: sources/dynamodb/streaming.go [hunk 2]
```diff
@@ -58,3 +68,186 @@ func NewDynamoDBStream(client dynamodbiface.DynamoDBAPI, srcTable string) (strin
...
+
+	shardId := *shard.ShardId
+	streamInfo.SetShardStatus(shardId, false)
+
+	var lastEvaluatedSequenceNumber *string = nil
+	passAfterUserExit := false
+	for {
+		shardIterator, err := getShardIterator(streamClient, lastEvaluatedSequenceNumber, shardId, streamArn)
+		if err != nil {
+			if CheckTrimmedDataError(err) {
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:152]: Add retries for expired shards

**Review by shreyakhajanchi — COMMENTED** [review]
Added minor comments. Overall looks good

File: sources/dynamodb/streaming_info.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// TotalUnexpecteds returns the total number of distinct unexpected conditions
+// encountered during processing of DynamoDB Streams.
+func (info *Info) TotalUnexpecteds() int64 {
+	return int64(len(info.Unexpecteds))
+}
+
+// Records returns the total number of records read from DynamoDB Streams.
+func (info *Info) TotalRecords() int64 {
```
[the remark this answers is not in the document]
Comment (nareshz) [reply sources/dynamodb/streaming_info.go:87]: This may be used at the time of printing reports. For the moment, I will remove it because there is no surety whether this will be used or not.

File: sources/dynamodb/streaming_info_test.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package dynamodb
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming_info_test.go:21]: I think it should be feasible to add test methods for all the methods defined in streaming_info.go

File: sources/dynamodb/streaming.go [hunk 5]
```diff
@@ -58,3 +68,190 @@ func NewDynamoDBStream(client dynamodbiface.DynamoDBAPI, srcTable string) (strin
...
+				continue
+			} else {
+				streamInfo.Unexpected(fmt.Sprintf("Couldn't get shardIterator for table %s: %s", srcTable, err))
+				break
+			}
+		}
+
+		getRecordsOutput, err := getRecords(streamClient, shardIterator)
+		if err != nil {
+			if checkTrimmedDataError(err) && retryCount < 5 {
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:164]: Add a comment on the purpose of this retry logic

File: sources/dynamodb/streaming_info_test.go (hunk 4 above)
Comment (nareshz) [reply sources/dynamodb/streaming_info_test.go:21]: Okay, I will do that.

File: sources/dynamodb/streaming_info.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// makeRecordMaps initializes maps used to stores record count for
+// a given table.
+func (info *StreamingInfo) makeRecordMaps(srcTable string) {
+	info.Records[srcTable] = make(map[string]int64)
+}
+
+// SetShardStatus changes the processing status of a shard.
+//
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming_info.go:50]: this line of comment got deleted unintentionally i guess

**Review by shreyakhajanchi — APPROVED** [review]
LGTM

File: sources/dynamodb/streaming_info.go (hunk 6 above)
Comment (nareshz) [reply sources/dynamodb/streaming_info.go:50]: actually this // is used to break lines here. It is a kind of printing "\n"
